File: Final/Code/Project_4/BaiToan5/Ex1_7/ex1_7.cpp
#include <iostream>
#include <queue>

struct Node {
    int val;
    Node* left;
    Node* right;

    Node(int v) : val(v), left(nullptr), right(nullptr) {}
};

class CompleteBinaryTree {
private:
    Node* root;

public:
    CompleteBinaryTree() : root(nullptr) {}

    void buildTree(int n) {
        if (n == 0) return;
        root = new Node(1);
        std::queue<Node*> q;
        q.push(root);
        int count = 1;

        while (count < n) {
            Node* current = q.front();
            q.pop();

            if (count < n) {
                current->left = new Node(++count);
                q.push(current->left);
            }
            if (count < n) {
                current->right = new Node(++count);
                q.push(current->right);
            }
        }
    }

    void printLevelOrder() {
        if (!root) return;
        std::queue<Node*> q;
        q.push(root);

        while (!q.empty()) {
            Node* current = q.front();
            q.pop();
            std::cout << current->val << " ";
            if (current->left) q.push(current->left);
            if (current->right) q.push(current->right);
        }
        std::cout << "\n";
    }
};

int main() {
    int n;
    std::cout << "Nhập số nút n: ";
    std::cin >> n;

    CompleteBinaryTree tree;
    tree.buildTree(n);

    std::cout << "Cây nhị phân đầy đủ (duyệt theo tầng):\n";
    tree.printLevelOrder();

    return 0;
}

File: Final/Code/Project_4/BaiToan5/Ex1_7/ex1_7.py
from collections import deque

class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

class CompleteBinaryTree:
    def __init__(self):
        self.root = None
        self.count = 0

    def add_node(self, val):
        new_node = Node(val)
        if not self.root:
            self.root = new_node
            return new_node
        
        q = deque()
        q.append(self.root)

        while q:
            node = q.popleft()
            if not node.left:
                node.left = new_node
                return new_node
            elif not node.right:
                node.right = new_node
                return new_node
            else:
                q.append(node.left)
                q.append(node.right)

    def build_tree(self, n):
        for i in range(1, n + 1):
            self.add_node(i)

    def print_level_order(self):
        if not self.root:
            return
        q = deque()
        q.append(self.root)
        while q:
            node = q.popleft()
            print(node.val, end=' ')
            if node.left:
                q.append(node.left)
            if node.right:
                q.append(node.right)

n = int(input("Nhập số nút n: "))
tree = CompleteBinaryTree()
tree.build_tree(n)
print("Cây nhị phân đầy đủ (duyệt theo tầng):")
tree.print_level_order()